cupyx/scipy/sparse/index.py

```diff
@@ -0,0 +1,854 @@
+"""Indexing mixin for sparse matrix classes.
+"""
+import math
+from .sputils import isintlike
+
+from numpy import integer
+from numpy import dtype
+
+import cupy
+from cupy import core
+
+
+INT_TYPES = (int, integer)
+
+int32_dtype = dtype('int32')
+int64_dtype = dtype('int64')
+float32_dtype = dtype('float32')
+float64_dtype = dtype('float64')
+complex64_dtype = dtype('complex64')
+complex128_dtype = dtype('complex128')
+
+module_options = ('--std=c++11',)
+
+
+def _broadcast_arrays(a, b):
+    """
+    Same as cupy.broadcast_arrays(a, b) but old writeability rules.
+    NumPy >= 1.17.0 transitions broadcast_arrays to return
+    read-only arrays. Set writeability explicitly to avoid warnings.
+    Retain the old writeability rules, as our Cython code assumes
+    the old behavior.
+    """
+    x, y = cupy.broadcast_arrays(a, b)
+
+    # Writeable doesn't seem to exist on flags in cupy
+    # x.flags.writeable = a.flags.writeable
+    # y.flags.writeable = b.flags.writeable
+    return x, y
+
+
+bin_col_offsets_ker_types = {dtype('int32'): 'bin_col_offsets_ker_str<int>'}
+bin_col_offsets_ker = core.RawModule(code="""
+    template<typename I>
+    __global__ void bin_col_offsets_ker_str(I n_idx,
+                                            I *col_idxs,
+                                            I *col_offsets) {
+
+        // Get the index of the thread
+        I jj = blockIdx.x * blockDim.x + threadIdx.x;
+
+        if(jj < n_idx) {
+            const I j = col_idxs[jj];
+            atomicAdd(col_offsets+j, 1);
+        }
+    }
+""", options=module_options, name_expressions=tuple(
+    bin_col_offsets_ker_types.values()))
+
+
+def bin_col_offsets(n_idx, col_ids, col_offsets, tpb=32):
+    grid = math.ceil(n_idx / tpb)
+    kernel = bin_col_offsets_ker.get_function(
+        bin_col_offsets_ker_types[int32_dtype])
+    kernel((grid,), (tpb,), (n_idx, col_ids, col_offsets))
+
+
+csr_column_index1_ker_types = {int32_dtype: 'csr_column_index1_ker<int>'}
+csr_column_index1_ker = core.RawModule(code="""
```

Comment:
It looks that this kernel can be more parallelized by using advanced indexing and cumsum.

Reply:
I’d like to try this but I’m having trouble thinking through how advanced indexing can be used efficiently here. Can you elaborate on your thoughts here?

Reply:
For example, this kernel can be written as follows:
```py
def _csr_column_index1_degree(n_row, col_offsets, Ap, Aj, Bp):
    col_sum = cupy.zeros((Aj.size + 1,), dtype=col_offsets.dtype)
    cupy.cumsum(col_offsets[Aj], out=col_sum[1:])
    Bp = col_sum[Ap]
    Bp[1:] -= Bp[:-1]
    return Bp
```